src/modules/kvs/commit.c

```diff
@@ -333,62 +346,83 @@ static int commit_link_dirent (commit_t *c, int current_epoch,
     }
 
     /* This is the first part of a key with multiple path components.
-     * Make sure that it is a directory in DIRVAL form, then recurse
-     * on the remaining path components.
+     * Make sure that it is a treeobj dir, then recurse on the
+     * remaining path components.
      */
     while ((next = strchr (name, '.'))) {
         *next++ = '\0';
-        if (!(subdirent = json_object_get (dir, name))) {
+
+        if (!treeobj_is_dir (dir)) {
+            saved_errno = EPERM;
+            goto done;
+        }
+
+        if (!(dir_entry = treeobj_get_entry (dir, name))) {
             if (json_is_null (dirent)) /* key deletion - it doesn't exist so return */
                 goto success;
-            if (!(subdir = json_object ())) {
-                saved_errno = ENOMEM;
+            if (!(subdir = treeobj_create_dir ())) {
+                saved_errno = errno;
                 goto done;
             }
-            if (!(tmpdirent = j_dirent_create ("DIRVAL", subdir))) {
+            if (treeobj_insert_entry (dir, name, subdir) < 0) {
                 saved_errno = errno;
                 json_decref (subdir);
                 goto done;
             }
-            if (json_object_set_new (dir, name, tmpdirent) < 0) {
-                json_decref (tmpdirent);
-                json_decref (subdir);
-                saved_errno = ENOMEM;
+            json_decref (subdir);
+        } else if (treeobj_is_dir (dir_entry)) {
+            subdir = dir_entry;
+        } else if (treeobj_is_dirref (dir_entry)) {
+            const char *ref;
+            int refcount;
+
+            if ((refcount = treeobj_get_count (dir_entry)) < 0) {
+                saved_errno = errno;
                 goto done;
             }
-            json_decref (subdir);
-        } else if ((o = json_object_get (subdirent, "DIRVAL"))) {
-            subdir = o;
-        } else if ((o = json_object_get (subdirent, "DIRREF"))) {
-            assert (json_is_string (o));
-            const char *ref = json_string_value (o);
+
+            if (refcount != 1) {
+                saved_errno = EPERM;
+                goto done;
+            }
+
+            if (!(ref = treeobj_get_blobref (dir_entry, 0))) {
+                saved_errno = errno;
+                goto done;
+            }
+
             if (!(subdir = cache_lookup_and_get_json (c->cm->cache,
                                                       ref,
                                                       current_epoch))) {
                 *missing_ref = ref;
                 goto success; /* stall */
             }
-            /* do not corrupt store by modify orig. */
-            if (!(subdir = json_copy (subdir))) {
+
+            /* do not corrupt store by modifying orig. */
+            if (!(subdir = treeobj_copy_dir (subdir))) {
                 saved_errno = ENOMEM;
```

Comment:
treeobj_copy_dir() can return other errors.  If all others are not possible due to previous checks, then perhaps an assert is appropriate here.